src/Type/Php/JsonDecodeDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,108 @@
+<?php declare(strict_types=1);
+
+
+namespace TheCodingMachine\Safe\PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\BitwiseFlagHelper;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\ConstantScalarType;
+use PHPStan\Type\ConstantTypeHelper;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NeverType;
+use PHPStan\Type\ObjectWithoutClassType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use Safe\Exceptions\JsonException;
+
+/**
+ * @see \PHPStan\Type\Php\JsonThrowOnErrorDynamicReturnTypeExtension
+ */
+final class JsonDecodeDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+    public function __construct(
+        private readonly BitwiseFlagHelper $bitwiseFlagAnalyser,
```

Comment:
instead of copying the whole extension you could instead try to inject `JsonThrowOnErrorDynamicReturnTypeExtension` here and delegate in `getTypeFromFunctionCall` the call to it.

that way this extensions reflects 1:1 whatever the PHPStan extesion in the phpstan-src repo will evolve to

Reply:
However do you think I could separate `JsonThrowOnErrorDynamicReturnTypeExtension` into two different extensions, one for `json_encode` and one for `json_decode`? As far as I can tell, they two code paths do not really  share any code.

Reply:
couldn't you ask the `ReflectionProvider` for the `json_decode` reflection and pass this along instead of the $functionReflection this extension gets passed into?

Reply:
I did not know that was possible. :+1: